controllers/databaseengine_controller.go

```diff
@@ -103,41 +104,34 @@ func (r *DatabaseEngineReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	}
 	dbEngine.Status.PendingOperatorUpgrades = pendingUpgrades
 
-	requeue := false
-	if done, err := r.handleOperatorUpgrade(ctx, dbEngine); err != nil {
-		if !errors.Is(err, errInstallPlanNotFound) {
-			return ctrl.Result{}, err
-		}
-		// We could not find the InstallPlan for the operator upgrade,
-		// so we will fallthrough since we'd still want the engine status to be updated.
-		// We will still requeue to check for the InstallPlan later.
-		requeue = true
-		logger.Error(err, "Upgrade failed, cannot find InstallPlan")
-	} else if !done {
-		// Upgrade is not complete, we will update the status and requeue.
-		return ctrl.Result{RequeueAfter: requeueAfter},
-			r.Status().Update(ctx, dbEngine)
-	}
-
 	dbEngine.Status.State = everestv1alpha1.DBEngineStateNotInstalled
-	dbEngine.Status.OperatorVersion = ""
 	ready, version, err := r.getOperatorStatus(ctx, req.NamespacedName)
-	if client.IgnoreNotFound(err) != nil {
+	if err != nil {
+		// Operator not installed, upgrade the status and return.
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{},
+				r.Status().Update(ctx, dbEngine)
+		}
 		return ctrl.Result{}, err
 	}
 
-	if version != "" {
-		dbEngine.Status.OperatorVersion = version
-		dbEngine.Status.State = everestv1alpha1.DBEngineStateInstalling
-	}
-
 	// Not ready yet, update status and check again later.
 	if !ready {
-		return ctrl.Result{RequeueAfter: requeueAfter},
+		dbEngine.Status.State = everestv1alpha1.DBEngineStateInstalling
+		if err := r.reconcileOperatorUpgradeStatus(ctx, dbEngine); err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{},
 			r.Status().Update(ctx, dbEngine)
 	}
 
+	dbEngine.Status.OperatorVersion = version
+	if err := r.tryUnlockDBEngine(ctx, dbEngine); err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
Will this actually have the intended effect? If the annotation is set and nothing happens the reconciliation will never get triggered and the lock will stay locked. Should we somehow re-queue a reconciliation to the time when the unlocking should have already happened?

Reply:
Hmm, fair point, although Kubernetes still does a period requeue regardless. But, still good to explicitly requeue to prevent unexpected delays

Reply:
Done